include/envoy/upstream/host_description.h

```diff
@@ -26,14 +28,24 @@ namespace Upstream {
   COUNTER(rq_success)                                                                              \
   COUNTER(rq_timeout)                                                                              \
   COUNTER(rq_total)                                                                                \
-  GAUGE(cx_active, Accumulate)                                                                     \
-  GAUGE(rq_active, Accumulate)
+  GAUGE(cx_active)                                                                                 \
+  GAUGE(rq_active)
 
 /**
  * All per host stats defined. @see stats_macros.h
  */
 struct HostStats {
-  ALL_HOST_STATS(GENERATE_COUNTER_STRUCT, GENERATE_GAUGE_STRUCT)
+  ALL_HOST_STATS(GENERATE_PRIMITIVE_COUNTER_STRUCT, GENERATE_PRIMITIVE_GAUGE_STRUCT);
+
+  // Provide access to counters as a map.
+  std::map<std::string, std::reference_wrapper<const Stats::PrimitiveCounter>> counters() const {
+    return {ALL_HOST_STATS(PRIMITIVE_COUNTER_NAME_AND_REFERENCE, IGNORE_PRIMITIVE_GAUGE)};
```

Comment:
This is a pretty cool hack; IIUC you are not storing the names anywhere in the heap, but have access to them only as temps.

This will improve steady-state memory usage. However, I think whenever the admin /stats endpoint is ever called, this will spike memory usage (by how much I do not know), as the map<string, ref> is perhaps several times less efficient than a vector of stat-refs with the name held in the stat.

Unfortunately the stats-memory integration test doesn't currently measure that temporary spike in memory highwater.

WDYT of changing the API here to call a lambda for every name/stat pair, with the name passed as a string_view? Then the callee could collect them as pairs and sort them? Also in my experience it's the same time-complexity but a constant-factor better to collect a vector and sort it than collect a map and iterate over it, for various reasons including the nodes/pointers overhead of the rb-tree underlying std::map. This is probably second order though; the larger question in my mind is the burst of memory overhead.

Reply:
The maps in question have at most 6 elements in them, so memory usage shouldn't spike.  The main reason why I opted for a map here was to avoid forcing the caller to do a sort.  Returning a std::vector<std::pair<std::string_view, std::reference_wrapper<const Stats::PrimitiveCounter>>> is possible.  Should I switch to that API instead?

Reply:
Sure that sounds good; you can just sort first before emitting. Up to you if you want to return as the vector or call a lambda; looks like either would be straightforward at the call-site.

Reply:
Actually you control the input to this -- the stat macro calls -- so you can just sort them in the declaration.

Reply:
The challenge may be in ensuring that they remain sorted.  But that's something I could write a test for.  Sorting the declaration seems fine to me, I'll try that out.

Reply:
Changed return type and added test to verify counter and gauge names are sorted